File: Entities/Order/Customer.cs
using Bogus;
using Bogus.Extensions;
using System;
using System.Collections.Generic;
using System.Text;

namespace owl_shop.Entities.Order
{
	class Customer
	{
		public string ID { get; set; }
		public string FirstName { get; set; }
		public string LastName { get; set; }
		public string? CompanyName { get; set; }
		public string Email { get; set; }
		public CustomerType Type { get; set; }

		public static Customer CreateRandom()
		{
			return new Faker<Customer>()
				.StrictMode(true)
				.RuleFor(c => c.ID, f => f.Random.Uuid().ToString())
				.RuleFor(c => c.FirstName, f => f.Person.FirstName)
				.RuleFor(c => c.LastName, f => f.Person.LastName)
				.RuleFor(c => c.CompanyName, f => f.Company.CompanyName().OrNull(f, .97f))
				.RuleFor(c => c.Email, (f, c) => f.Internet.Email(c.FirstName, c.LastName))
				.RuleFor(c => c.Type, f => f.Random.Enum<CustomerType>())
				.Generate();
		}
	}

	enum CustomerType
	{
		Personal,
		Business,
		NonProfit,
	}
}

File: pkg/fake/customer.go
package fake

import (
	"github.com/brianvoe/gofakeit/v5"
	"github.com/mroth/weightedrand"

	shoppb "github.com/cloudhut/owl-shop/pkg/protogen/shop/v1"
)

type CustomerType string

const (
	CustomerTypePersonal CustomerType = "PERSONAL"
	CustomerTypeBusiness CustomerType = "BUSINESS"
)

type Customer struct {
	// VersionedStruct
	Version int `json:"version"`

	ID           string       `json:"id"`
	FirstName    string       `json:"firstName"`
	LastName     string       `json:"lastName"`
	Gender       string       `json:"gender"`
	CompanyName  *string      `json:"companyName"`
	Email        string       `json:"email"`
	CustomerType CustomerType `json:"customerType"` // PERSONAL | BUSINESS
	Revision     int          `json:"revision"`     // Each change on the customer increments the revision
}

func (c *Customer) Protobuf() *shoppb.Customer {
	companyName := ""
	if c.CompanyName != nil {
		companyName = *c.CompanyName
	}

	customerType := shoppb.Customer_CUSTOMER_TYPE_PERSONAL
	if c.CustomerType == CustomerTypeBusiness {
		customerType = shoppb.Customer_CUSTOMER_TYPE_BUSINESS
	}

	return &shoppb.Customer{
		Version:      int32(c.Version),
		Id:           c.ID,
		FirstName:    c.FirstName,
		LastName:     c.LastName,
		Gender:       c.Gender,
		CompanyName:  companyName,
		Email:        c.Email,
		CustomerType: customerType,
		Revision:     int32(c.Revision),
	}
}

func NewCustomer() Customer {
	person := gofakeit.Person()

	var companyName *string
	customerType := newCustomerType()
	if customerType == CustomerTypeBusiness {
		company := gofakeit.Company()
		companyName = &company
	}

	return Customer{
		Version:      0,
		ID:           gofakeit.UUID(),
		FirstName:    person.FirstName,
		LastName:     person.LastName,
		Gender:       person.Gender,
		CompanyName:  companyName,
		Email:        gofakeit.Email(),
		CustomerType: customerType,
	}
}

// newCustomerType returns a customer type based on a weighted random choice
func newCustomerType() CustomerType {
	c, err := weightedrand.NewChooser(
		weightedrand.Choice{Item: CustomerTypePersonal, Weight: 99},
		weightedrand.Choice{Item: CustomerTypeBusiness, Weight: 1},
	)
	if err != nil {
		panic(err)
	}
	customerType := c.Pick().(CustomerType)
	return customerType
}
